src/asm/main.c

```diff
@@ -6,815 +6,96 @@
-{
-	char			*label;
-	t_instruction	*ptr;
-
-	label = str;
-	if (is_direct_label(str))
-		label = &str[2];
-	ptr = a->instruction;
-	while (ptr != NULL)
-	{
-		if (ptr->type == T_LAB)
-		{
-			if (ft_strcmp(ptr->label, label) == 0)
-				return (ptr);
-		}
-		ptr = ptr->next;
-	}
-	return (NULL);
-}
-
-void	resolve_labels(t_asm *a)
-{
-	t_instruction	*ptr;
-	t_instruction	*label;
-	int				i;
-	
-	ptr = a->instruction;
-	while (ptr != NULL)
-	{
-		if (ptr->type == T_REG)
-		{
-			i = 0;
-			while (i < ptr->arg_count)
-			{
-				if (is_direct_label(ptr->operands[i]))
-				{
-					label = find_label(a, ptr->operands[i]);
-					if (label == NULL)
-						ft_printf("error: label not found '%s'\n", ptr->operands[i]);
-					else
-						ptr->op_values[i] = label->address - ptr->address;
-				}
-				else if (is_register(ptr->operands[i]))
-					ptr->op_values[i] = ft_atoi(&ptr->operands[i][1]);
-				else if (is_indirect(ptr->operands[i]))
-					ptr->op_values[i] = ft_atoi(ptr->operands[i]);
-				else if (is_direct(ptr->operands[i]))
-					ptr->op_values[i] = ft_atoi(&ptr->operands[i][1]);
-				i++;
-			}
-		}
-		ptr = ptr->next;
-	}
-}
-
-void	generate_output(t_asm *a)
-{
-	int				i;
-	int				j;
-	t_instruction	*ptr;
-
-	a->header.magic = betole(COREWAR_EXEC_MAGIC);
-	a->header.prog_size = betole(a->prog_size);
-	ptr = a->instruction;
-	i = 0;
-	while (ptr)
-	{
-		if (ptr->type == T_REG)
-		{
-			a->body[i++] = (char)ptr->opcode;
-			if (ptr->has_acb)
-				a->body[i++] = (char)ptr->acb;
-			j = 0;
-			while (j < ptr->arg_count)
-			{
-				if (ptr->op_types[j] == T_REG)
-					a->body[i++] = (char)ptr->op_values[j];
-				if ((ptr->op_types[j] == T_IND) || ((ptr->op_types[j] == T_DIR) && (ptr->short_direct == 1)))
-				{
-					a->body[i++] = (char)(ptr->op_values[j] >> 8);
-					a->body[i++] = (char)ptr->op_values[j];
-				}
-				if ((ptr->op_types[j] == T_DIR) && (ptr->short_direct == 0))
-				{
-					a->body[i++] = (char)(ptr->op_values[j] >> 24);
-					a->body[i++] = (char)(ptr->op_values[j] >> 16);
-					a->body[i++] = (char)(ptr->op_values[j] >> 8);
-					a->body[i++] = (char)ptr->op_values[j];
-				}
-				j++;
-			}
-		}
-		ptr = ptr->next;
-	}
-}
-
-void	routine(t_asm *a)
-{
-	get_mandatory_info(a);
-	parse_instructions(a);
-	resolve_labels(a);
-	ft_printf("prog size: %d\n", a->prog_size);
-	if (a->prog_size > CHAMP_MAX_SIZE)
-		ft_printf("error: prog size exceeds CHAMP_MAX_SIZE\n");
-	generate_output(a);
-	//
-	//destroy instruction linked list
-	//
-	put_file_contents("assembler.output", (unsigned char *)&a->header, sizeof(t_header) + a->prog_size);
-
-
-	destroy_char_array(a->split, a->lines);
-	free(a);
-}
-
 int		main(int ac, char **av)
 {
-	int		fd;
-	t_asm	*a;
-
 	if (ac == 2)
+		init(av, 1, 0);
+	else if (ac == 3)
 	{
-		if ((fd = open(av[1], O_RDONLY)) != -1)
-		{
-			a = ft_memalloc(sizeof(t_asm));
-			a->fd = fd;
-			a->filename = av[1];
-			routine(a);
-			close(fd);
-			return (0);
-		}
+		if (ft_strcmp(av[1], "-a") == 0)
```

Comment:
I'd prefer the usage of a real options parser, but this is okay for now.

Reply:
Options parser is going to need to be written regardless (for parsing vm arguments), which I can work on later which can be ported to assembler.